File: src/main/java/com/gitlab/daring/image/swing/BaseAction.java
package com.gitlab.daring.image.swing;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.util.function.Consumer;

import static javax.swing.JComponent.WHEN_IN_FOCUSED_WINDOW;
import static javax.swing.KeyStroke.getKeyStroke;

public class BaseAction extends AbstractAction {

    final Consumer<ActionEvent> act;

    public BaseAction(String name, Consumer<ActionEvent> act) {
        super(name);
        this.act = act;
    }

    public String getName() {
        return (String) getValue(NAME);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        act.accept(e);
    }

    public void register(JComponent c, String keyStr) {
        c.getActionMap().put(getName(), this);
        if (!keyStr.isEmpty()) {
            KeyStroke ks = getKeyStroke(keyStr);
            c.getInputMap(WHEN_IN_FOCUSED_WINDOW).put(ks, getName());
        }
    }

}

File: src/main/kotlin/com/gitlab/daring/image/swing/BaseAction.kt
package com.gitlab.daring.image.swing

import java.awt.event.ActionEvent
import javax.swing.AbstractAction
import javax.swing.JComponent
import javax.swing.JComponent.WHEN_FOCUSED
import javax.swing.JComponent.WHEN_IN_FOCUSED_WINDOW
import javax.swing.KeyStroke.getKeyStroke

class BaseAction(name: String, val act: () -> Unit) : AbstractAction(name) {

    constructor(name: String, act: Runnable) : this(name, act::run)

    val name get() = getValue(NAME) as String

    override fun actionPerformed(e: ActionEvent) {
        act.invoke()
    }

    fun register(c: JComponent, keyStr: String, inWindow: Boolean = false) {
        c.actionMap.put(name, this)
        if (!keyStr.isEmpty()) {
            val ks = getKeyStroke(keyStr)
            val cond = if (inWindow) WHEN_IN_FOCUSED_WINDOW else WHEN_FOCUSED
            c.getInputMap(cond).put(ks, name)
        }
    }

}